src/lib.rs

```diff
@@ -53,9 +54,90 @@ fn it_works() {
         }
     }
     let mut foo = Foo::A;
-    take(&mut foo, |mut f| {
+    take(&mut foo, |f| {
        drop(f);
        Foo::B
     });
     assert_eq!(&foo, &Foo::B);
-}
\ No newline at end of file
+}
+
+/// Allows use of a value pointed to by `&mut T` as though it was owned, as long as a `T` is made available afterwards.
+///
+/// The closure must return a valid T.
+/// # Important
+/// Will replace `&mut T` with `recover` if the closure panics, then continues the panic.
+///
+/// # Example
+/// ```
+/// struct Foo;
+/// let mut foo = Foo;
+/// take_mut::take_or_recover(&mut foo, Foo, |foo| {
+///     // Can now consume the Foo, and provide a new value later
+///     drop(foo);
+///     // Do more stuff
+///     Foo // Return new Foo from closure, which goes back into the &mut Foo
+/// });
+/// ```
+pub fn take_or_recover<T, F>(mut_ref: &mut T, recover: T, closure: F)
+  where F: FnOnce(T) -> T {
+    use std::ptr;
+    unsafe {
+        let old_t = ptr::read(mut_ref);
+        let new_t = panic::catch_unwind(panic::AssertUnwindSafe(|| closure(old_t)));
```

Comment:
Using AssertUnwindSafe here for the whole closure and old_t should be safe as we just intercept the panic, write our pointer (to the mutable reference that can't be modified by the closure in safe code!) and then pass on the panic.

Catching the panic in a recovering way without passing it onwards and the safeness involved with that would be the job of the caller if it wishes so, just like in cases without usage of take_or_recover()